File: LeetCode/1371. 每个元音包含偶数次的最长子字符串.cpp
class Solution
{
public:
  int findTheLongestSubstring(string s)
  {
    vector<int> statusPos(1 << 5, -1);
    int ans = 0, status = 0;
    statusPos[0] = 0;
    for (int i = 0; i < s.length(); i++)
    {
      if (s[i] == 'a')
        status ^= 1 << 0;
      else if (s[i] == 'e')
        status ^= 1 << 1;
      else if (s[i] == 'i')
        status ^= 1 << 2;
      else if (s[i] == 'o')
        status ^= 1 << 3;
      else if (s[i] == 'u')
        status ^= 1 << 4;

      if (~statusPos[status])
        ans = max(ans, i + 1 - statusPos[status]);
      else
        statusPos[status] = i + 1;
    }
    return ans;
  }
};

File: LeetCode/1371. 每个元音包含偶数次的最长子字符串.rs
use std::cmp::max;
impl Solution {
    pub fn find_the_longest_substring(s: String) -> i32 {
        let mut status_pos:Vec<i32>=vec![-1;1<<5];
        let (mut status,mut ans)=(0,0);
        status_pos[0]=0;
        for (i,c) in s.char_indices(){
            match c{
                'a'=>status^=1<<0,
                'e'=>status^=1<<1,
                'i'=>status^=1<<2,
                'o'=>status^=1<<3,
                'u'=>status^=1<<4,
                _=>()
            }
            match status_pos[status]{
                -1=>status_pos[status]=(i as i32)+1,
                _=>ans=max(ans,(i as i32)+1-status_pos[status])
            }
        }
        ans
    }
}